Visual Studio 2022 17.7 reformat on save isn't working

I installed the Visual Studio 2022 extension, and I just can't get it to work:
- I enabled "Reformat on save" and no reformatting happens on save
- When I right-click, the menu option to reformat is always disabled. (See below)

(Yes, I did run "dotnet tool install -g csharpier", and yes, the tool does run from the command-line.)

![image](https://user-images.githubusercontent.com/78368152/212428260-62a549f6-07dc-4c59-b90e-10397aeae49b.png)

Any suggestions to diagnose what's wrong?

I have a very similar problem.

I did not have any Csharpier folders in appdata.
When i check the Output log for for Cshapier, after clicking install i get
`["INFO" - 10:52:43 AM] Initializing csharpier-vscode
`

and nothing else. Then when I open a Text file or try to format a C# text file:
`["ERROR" - 10:54:19 AM] CSharpier was not found so files may not be formatted.
`

and the popup about _CSharpier needs to be installed to support formatting files_
When i clock install globally or locally:
`["INFO" - 10:55:23 AM] Installing csharpier globally with dotnet tool install -g csharpier
`

I tried running several commands from Powershell:
`dotnet tool install -g csharpier`
`PS C:\Users\jha> dotnet tool install csharpier --version 0.24.2 --tool-path C:\Users\jha\AppData\Local\CSharpier\0.24.2`

I get the following error:
```
C:\Users\jha\AppData\Local\Temp\2286cf84-ffa9-4603-8e6b-8e481e2097eb\restore.csproj : error NU1100: Unable to resolve '
csharpier (>= 0.0.0)' for 'net6.0'.
C:\Users\jha\AppData\Local\Temp\2286cf84-ffa9-4603-8e6b-8e481e2097eb\restore.csproj : error NU1100: Unable to resolve '
csharpier (>= 0.0.0)' for 'net6.0/any'.
The tool package could not be restored.
Tool 'csharpier' failed to install. This failure may have been caused by:

* You are attempting to install a preview release and did not use the --version option to specify the version.
* A package by this name was found, but it was not a .NET tool.
* The required NuGet feed cannot be accessed, perhaps because of an Internet connection problem.
* You mistyped the name of the tool.

For more reasons, including package naming enforcement, visit https://aka.ms/failure-installing-tool
```

I have tried both not6.0 and net7.0.
I have tried reinstalling VS code, Csharpier, .net but to no avail.

--- a/Src/CSharpier.VisualStudio/CSharpier.VisualStudio/CSharpier.VisualStudio.csproj
+++ b/Src/CSharpier.VisualStudio/CSharpier.VisualStudio/CSharpier.VisualStudio.csproj
@@ -75,8 +75,11 @@
     <Reference Include="System.Xml" />
   </ItemGroup>
   <ItemGroup>
-    <PackageReference Include="Microsoft.VisualStudio.SDK" Version="17.0.31902.203" ExcludeAssets="runtime" />
-    <PackageReference Include="Microsoft.VSSDK.BuildTools" Version="17.0.5233" />
+    <PackageReference Include="Microsoft.VisualStudio.SDK" Version="17.6.36389" ExcludeAssets="runtime" />
+    <PackageReference Include="Microsoft.VSSDK.BuildTools" Version="17.6.2164">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
   </ItemGroup>
   <Import Project="..\CSharpier.VisualStudioShared\CSharpier.VisualStudioShared.projitems" Label="Shared" />
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />

--- a/Src/CSharpier.VisualStudio/CSharpier.VisualStudio/source.extension.vsixmanifest
+++ b/Src/CSharpier.VisualStudio/CSharpier.VisualStudio/source.extension.vsixmanifest
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
     <Metadata>
-        <Identity Id="83d6b6a0-9e25-4034-80f3-38445d8a8837" Version="1.4.2" Language="en-US" Publisher="CSharpier" />
+        <Identity Id="83d6b6a0-9e25-4034-80f3-38445d8a8837" Version="1.4.3" Language="en-US" Publisher="CSharpier" />
         <DisplayName>CSharpier</DisplayName>
         <Description xml:space="preserve">CSharpier is an opinionated code formatter for c#. It uses Roslyn to parse your code and re-prints it using its own rules.</Description>
         <MoreInfo>https://github.com/belav/csharpier</MoreInfo>

--- a/Src/CSharpier.VisualStudio/CSharpier.VisualStudio2019/CSharpier.VisualStudio2019.csproj
+++ b/Src/CSharpier.VisualStudio/CSharpier.VisualStudio2019/CSharpier.VisualStudio2019.csproj
@@ -34,7 +34,7 @@
     <DebugType>full</DebugType>
     <Optimize>false</Optimize>
     <OutputPath>bin\Debug\</OutputPath>
-    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <DefineConstants>TRACE;DEBUG;Dev16</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
@@ -42,7 +42,7 @@
     <DebugType>pdbonly</DebugType>
     <Optimize>true</Optimize>
     <OutputPath>bin\Release\</OutputPath>
-    <DefineConstants>TRACE</DefineConstants>
+    <DefineConstants>TRACE;Dev16</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
@@ -75,7 +75,7 @@
     <Reference Include="System.Xml" />
   </ItemGroup>
   <ItemGroup>
-  <PackageReference Include="Microsoft.VisualStudio.SDK" Version="16.0.206" ExcludeAssets="runtime" />
+    <PackageReference Include="Microsoft.VisualStudio.SDK" Version="16.0.206" ExcludeAssets="runtime" />
     <PackageReference Include="Microsoft.VSSDK.BuildTools" Version="16.11.35" />
   </ItemGroup>
   <ItemGroup>

--- a/Src/CSharpier.VisualStudio/CSharpier.VisualStudio2019/CSharpier.VisualStudio2019.csproj
+++ b/Src/CSharpier.VisualStudio/CSharpier.VisualStudio2019/CSharpier.VisualStudio2019.csproj
@@ -34,7 +34,7 @@
     <DebugType>full</DebugType>
     <Optimize>false</Optimize>
     <OutputPath>bin\Debug\</OutputPath>
-    <DefineConstants>TRACE;DEBUG;Dev16</DefineConstants>
+    <DefineConstants>TRACE;DEBUG;DEV16</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
@@ -42,7 +42,7 @@
     <DebugType>pdbonly</DebugType>
     <Optimize>true</Optimize>
     <OutputPath>bin\Release\</OutputPath>
-    <DefineConstants>TRACE;Dev16</DefineConstants>
+    <DefineConstants>TRACE;DEV16</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>

--- a/Src/CSharpier.VisualStudio/CSharpier.VisualStudio2019/source.extension.vsixmanifest
+++ b/Src/CSharpier.VisualStudio/CSharpier.VisualStudio2019/source.extension.vsixmanifest
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
     <Metadata>
-        <Identity Id="edd8b38c-baa1-46c6-b82e-1da7a0ba597b" Version="1.4.2" Language="en-US" Publisher="CSharpier" />
+        <Identity Id="edd8b38c-baa1-46c6-b82e-1da7a0ba597b" Version="1.4.3" Language="en-US" Publisher="CSharpier" />
         <DisplayName>CSharpier 2019</DisplayName>
         <Description xml:space="preserve">CSharpier is an opinionated code formatter for c#. It uses Roslyn to parse your code and re-prints it using its own rules.</Description>
         <MoreInfo>https://github.com/belav/csharpier</MoreInfo>

--- a/Src/CSharpier.VisualStudio/CSharpier.VisualStudioShared/ReformatWithCSharpierOnSave.cs
+++ b/Src/CSharpier.VisualStudio/CSharpier.VisualStudioShared/ReformatWithCSharpierOnSave.cs
@@ -8,6 +8,10 @@ using Task = System.Threading.Tasks.Task;
 namespace CSharpier.VisualStudio
 {
     public class ReformatWithCSharpierOnSave : IVsRunningDocTableEvents3
+#if !Dev16
+            , IVsRunningDocTableEvents7
+#endif
+
     {
         private readonly DTE dte;
         private readonly RunningDocumentTable runningDocumentTable;
@@ -28,6 +32,18 @@ namespace CSharpier.VisualStudio
         {
             var dte = await package.GetServiceAsync(typeof(DTE)) as DTE;
             new ReformatWithCSharpierOnSave(package, dte!);
+        }
+
+        public IVsTask OnBeforeSaveAsync(uint cookie, uint flags, IVsTask? saveTask)
+        {
+            return ThreadHelper.JoinableTaskFactory.RunAsyncAsVsTask(
+                VsTaskRunContext.UIThreadBackgroundPriority,
+                async cancellationToken =>
+                {
+                    await Task.Yield();
+                    return this.OnBeforeSave(cookie);
+                }
+            );
         }
 
         public int OnBeforeSave(uint docCookie)
@@ -103,6 +119,18 @@ namespace CSharpier.VisualStudio
         public int OnAfterSave(uint docCookie)
         {
             return VSConstants.S_OK;
+        }
+
+        public IVsTask? OnAfterSaveAsync(uint cookie, uint flags)
+        {
+            return ThreadHelper.JoinableTaskFactory.RunAsyncAsVsTask(
+                VsTaskRunContext.UIThreadBackgroundPriority,
+                async cancellationToken =>
+                {
+                    await Task.Yield();
+                    return VSConstants.S_OK;
+                }
+            );
         }
 
         public int OnAfterAttributeChange(uint docCookie, uint grfAttribs)

--- a/Src/CSharpier.VisualStudio/CSharpier.VisualStudioShared/ReformatWithCSharpierOnSave.cs
+++ b/Src/CSharpier.VisualStudio/CSharpier.VisualStudioShared/ReformatWithCSharpierOnSave.cs
@@ -8,7 +8,7 @@ using Task = System.Threading.Tasks.Task;
 namespace CSharpier.VisualStudio
 {
     public class ReformatWithCSharpierOnSave : IVsRunningDocTableEvents3
-#if !Dev16
+#if !DEV16
             , IVsRunningDocTableEvents7
 #endif
 
@@ -30,17 +30,18 @@ namespace CSharpier.VisualStudio
 
         public static async Task InitializeAsync(CSharpierPackage package)
         {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
             var dte = await package.GetServiceAsync(typeof(DTE)) as DTE;
             new ReformatWithCSharpierOnSave(package, dte!);
         }
 
-        public IVsTask OnBeforeSaveAsync(uint cookie, uint flags, IVsTask? saveTask)
+        public IVsTask? OnBeforeSaveAsync(uint cookie, uint flags, IVsTask? saveTask)
         {
             return ThreadHelper.JoinableTaskFactory.RunAsyncAsVsTask(
-                VsTaskRunContext.UIThreadBackgroundPriority,
-                async cancellationToken =>
+                VsTaskRunContext.UIThreadNormalPriority,
+                async _ =>
                 {
-                    await Task.Yield();
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
                     return this.OnBeforeSave(cookie);
                 }
             );
@@ -123,14 +124,7 @@ namespace CSharpier.VisualStudio
 
         public IVsTask? OnAfterSaveAsync(uint cookie, uint flags)
         {
-            return ThreadHelper.JoinableTaskFactory.RunAsyncAsVsTask(
-                VsTaskRunContext.UIThreadBackgroundPriority,
-                async cancellationToken =>
-                {
-                    await Task.Yield();
-                    return VSConstants.S_OK;
-                }
-            );
+            return null;
         }
 
         public int OnAfterAttributeChange(uint docCookie, uint grfAttribs)

--- a/Src/CSharpier.VisualStudio/ChangeLog.md
+++ b/Src/CSharpier.VisualStudio/ChangeLog.md
@@ -1,4 +1,7 @@
-﻿## [1.4.2]
+﻿## [1.4.3]
+- Fix for format on save sometimes not working in VS2022 17.7
+
+## [1.4.2]
 - Recompile with VS2022 17.7 preview 4 - fixes the format on save for 17.7 preview 4.
 
 ## [1.4.1]

